apps/settings/js/panels/operator_settings/model/auto_selection_model.js

```diff
@@ -0,0 +1,82 @@
+define(function() {
+  'use strict';
+
+  var Defer = require('modules/defer');
+  var FeatureModel = require('modules/feature_model');
+  var MobileConnectionWrapper =
+    require('panels/operator_settings/mobile_connection_wrapper');
+
+  // Auto selection state
+  const AS_STATE = {
+    UNKNOWN: 0,
+    ENABLED: 1,
+    DISABLED: 2
+  };
+
+  var AutoSelectionModel = function(connWrapper) {
+    return FeatureModel({
+      onInit: function(onchange) {
+        this._pendingState = null;
+        this._defer = null;
+        connWrapper.observe('state', (newState) => {
+          this._checkPendingState(newState);
+        });
+        return AS_STATE.ENABLED;
+      },
+      onGetState: function() {
+        switch (connWrapper.networkSelectionMode) {
+          case 'manual':
+            return Promise.resolve(AS_STATE.DISABLED);
+          case 'automatic':
+            return Promise.resolve(AS_STATE.ENABLED);
+          default:
+            return Promise.resolve(AS_STATE.UNKNOWN);
+        }
+      },
+      onSetState: function(state) {
+        switch (connWrapper.state) {
+          case MobileConnectionWrapper.STATE.IDLE:
+            return connWrapper.setAutoSelection().then(() => {
+              return AS_STATE.ENABLED;
+            }, () => {
+              return AS_STATE.DISABLED;
+            });
+          case MobileConnectionWrapper.STATE.BUSY:
+            this._pendingState = state;
+            if (!this._defer) {
+              this._defer = Defer();
+            }
+            return this._defer.promise;
+        }
+      },
+      _checkPendingState: function(newState) {
+        if (newState !== MobileConnectionWrapper.STATE.IDLE) {
+          return;
+        }
+        if (this._pendingState === AS_STATE.ENABLED) {
+          connWrapper.setAutoSelection().then(() => {
+            this._pendingState = null;
+            this._defer.resolve(AS_STATE.ENABLED);
+            this._defer = null;
+          }, () => {
+            this._pendingState = null;
+            this._defer.reject(AS_STATE.DISABLED);
+            this._defer = null;
+          });
+        } else if (this._pendingState === AS_STATE.DISABLED) {
+          this._pendingState = null;
+          this._defer.resolve(AS_STATE.DISABLED);
+          this._defer = null;
```

Comment:
this code snippet below is duplicated, maybe we can introduce another function to do some cleanup works here.

```javascript
this._pendingState = null;
this._defer = null;
```